### PR TITLE
API,Core: Add scan planning metrics for skipped data/delete files

### PR DESCRIPTION
**Pull request by nastra**

### CODE REVIEW

File: core/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java [hunk 1]
```diff
@@ -130,6 +130,71 @@ public void scanningWithDeletes() throws IOException {
...
+    assertThat(scanReport.scanMetrics().scannedDataManifests().value()).isEqualTo(1);
+    assertThat(scanReport.scanMetrics().skippedDataManifests().value()).isEqualTo(1);
+    assertThat(scanReport.scanMetrics().totalDataManifests().value()).isEqualTo(2);
+    assertThat(scanReport.scanMetrics().totalDeleteManifests().value()).isEqualTo(0);
+    assertThat(scanReport.scanMetrics().totalFileSizeInBytes().value()).isEqualTo(10L);
+    assertThat(scanReport.scanMetrics().totalDeleteFileSizeInBytes().value()).isEqualTo(0L);
+  }
+
+  @Test
+  public void scanningWithSkippedDeleteFiles() throws IOException {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java:167]: @rdblue is there a good/easy way to test scan planning with skipped delete files?

File: api/src/main/java/org/apache/iceberg/metrics/DefaultCounter.java [hunk 2]
```diff
@@ -38,6 +38,11 @@ public void increment(long amount) {}
         public long value() {
           throw new UnsupportedOperationException("NOOP counter has no value");
         }
+
+        @Override
+        public String toString() {
+          return "NOOP counter";
```
Comment (nastra) [inline api/src/main/java/org/apache/iceberg/metrics/DefaultCounter.java:44]: the default implementation of `toString()` would always use `value()` and show the `UnsupportedOperationException` in debugging sessions, thus we're just overriding this for Counter/Timer here

**Review by danielcweeks — APPROVED** [review]
+1 LGTM

**Comment by nastra** [comment]
Thanks for the review @danielcweeks. Let's wait with the merge until https://github.com/apache/iceberg/pull/5780 is in, because a lot of the visual noise in this diff will be gone then

**Comment by nastra** [comment]
@danielcweeks I've rebased this and it should be good to go

**Review by Fokko — APPROVED** [review]

File: core/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java [hunk 3]
```diff
@@ -61,16 +62,16 @@ public void scanningWithMultipleDataManifests() throws IOException {
...
+    ScanMetricsResult result = scanReport.scanMetrics();
+    assertThat(result.totalPlanningDuration().totalDuration()).isGreaterThan(Duration.ZERO);
+    assertThat(result.resultDataFiles().value()).isEqualTo(3);
+    assertThat(result.resultDeleteFiles().value()).isEqualTo(0);
+    assertThat(result.scannedDataManifests().value()).isEqualTo(2);
+    assertThat(result.skippedDataManifests().value()).isEqualTo(0);
+    assertThat(result.totalDataManifests().value()).isEqualTo(2);
+    assertThat(result.totalDeleteManifests().value()).isEqualTo(0);
+    assertThat(result.totalFileSizeInBytes().value()).isEqualTo(30L);
+    assertThat(result.totalDeleteFileSizeInBytes().value()).isEqualTo(0L);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java:74]: @nastra, why not add an assertion for the skipped manifest counters in existing tests?

Comment (nastra) [reply core/src/test/java/org/apache/iceberg/TestScanPlanningAndReporting.java:74]: I realized that those were missing in the existing tests when working on https://github.com/apache/iceberg/pull/5792 and included/added them there

**Pull request merged**